source/common/conn_pool/conn_pool_base.cc

```diff
@@ -608,11 +660,27 @@ void ConnPoolImplBase::onPendingStreamCancel(PendingStream& stream,
     state_.decrPendingStreams(1);
     stream.removeFromList(pending_streams_);
   }
-  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess && !connecting_clients_.empty() &&
-      connectingConnectionIsExcess()) {
-    auto& client = *connecting_clients_.front();
-    transitionActiveClientState(client, ActiveClient::State::Draining);
-    client.close();
+  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess) {
+    if (!connecting_clients_.empty() &&
+        connectingConnectionIsExcess(*connecting_clients_.front())) {
+      auto& client = *connecting_clients_.front();
+      transitionActiveClientState(client, ActiveClient::State::Draining);
+      client.close();
+    } else if (!early_data_clients_.empty()) {
+      ActiveClient* idle_client{nullptr};
+      for (ActiveClientPtr& client : early_data_clients_) {
+        if (client->numActiveStreams() == 0) {
+          if (connectingConnectionIsExcess(*client)) {
+            idle_client = client.get();
+          }
+          break;
+        }
+      }
+      if (idle_client != nullptr) {
```

Comment:
I think we could move this block up into the if () body on line 674 to slightly simplify.

Reply:
I don't think I understand. Which iterator will be messed up?  (It looks like we're breaking out of the iteration after we set idle_client so I would think it would be safe to use it there?)